dev_docs/contributing/kibana_http_api_design_guidelines.mdx

```diff
@@ -0,0 +1,613 @@
+
+`PUT /api/my_domain/my_resources/{id}` ✅
+
+`PUT /api/my_domain/my_resource/{id}` ❌
+
+**Prefix actions against resource collections with `_`**
+
+Sometimes we want to designate a variation or action against a resource:
+
+```
+POST /api/my_domain/my_resources/_bulk_delete
+```
+
+Carefully consider cases where this pattern may make sense as it is more user-friendly to expose simple paths and actions.
+
+### Methods
+
+**Side effects**
+
+* `GET` should not have side-effects
+* `PATCH`, `PUT` and `DELETE` should have side-effects (alter some resource)
+* `POST` should have side-effects (like creating some resource, read on!)
+
+**Same method, same result?**
+
+This is sometimes referred to as _idempotency_. It is related to side-effects, but not quite the same thing.
+
+Imagine we were to run the following methods 1000x over:
+
+* `GET` always results in the same _server state_
+* `PUT` always results in the same _server state_ (returns 409 after the first request)
+* `DELETE` always results in the same _server state_ (returns 404 after the first request)
+* `PATCH` always results in the same _server state_ for the same input! For example: `PATCH /api/coolstore/order/{id} { expected_delivery: "ASAP!" }` will result in the same state.
+
+However...
+
+* `POST` should result in a different _server state_. For example: `POST /api/coolstore/order` will result in a new order each time!
+
+**POST is (mostly) for creating a resource**
+
+Sometimes `POST` can be used for an action against a resource or for actions like a `GET` but with a body. In this case, `POST` may be both side-effect free and idempotent! In most cases it will not be because your request will open a PIT: see [pagination and sorting](#pagination-and-sorting).
+
+If you expect you need to support a lot of input for searching or listing resources it might be good to reserve `POST` for this purposes.
+
+**GET does not accept a body**
+
+`GET /api/my_domain/my_resources/{id}` ✅
+
+`GET /api/my_domain/my_resources/{id} { body: { params } }` ❌
+
+```
+POST /api/my_domain/my_resources
+{
+  search_after: "<a_really_long_id>",
+  size: 10000
+}
+```
+
+**PUT or PATCH are for updating a resource**
+
+`PUT` is a bit special in that it should _create_ the resource if it does not exist. It should expect the **full** resource in the body of the call.
+
+`PATCH` is used for partial updates and should never create a resource that does not already exist. It should expect a **partial** resource in the body of the call.
+
+**DELETE is for deleting a resource**
+
+Delete can return a simple response instead of the full resource:
+
+```
+DELETE /api/my_domain/my_resources/{id}
+# -> 204 No content
+```
+
+**Stick to simple methods**
+
+`GET`, `POST`, `PUT`, `PATCH`, `DELETE` should cover almost all cases. If you are considering `HEAD` or some other method make sure you have a good justification for your HTTP API.
+
+### Path and query parameters
+
+**Should use snake case**
```

Comment:
Is using snake case a suggestion or a must have?

"Should (as suggestion) use snake case" -> "Prefer snake case"

"Should (as must-have) use snake case" -> "Use snake case"

Reply:
We could leave this as it is for now and follow up with changing it to `Must use snake case`